File: app/src/main/java/com/jonlatane/beatpad/view/topology/TopologyView.java
package com.jonlatane.beatpad.view.topology;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.jonlatane.beatpad.R;
import com.jonlatane.beatpad.harmony.ChordSequence;
import com.jonlatane.beatpad.harmony.chord.Chord;

import java.util.ArrayList;
import java.util.List;

import static com.jonlatane.beatpad.harmony.ChordSequence.CHROMATIC;
import static com.jonlatane.beatpad.harmony.chord.Chord.MAJ_7;

/**
 * Created by jonlatane on 5/5/17.
 */
public class TopologyView extends RelativeLayout {
    private static final String TAG = TopologyView.class.getSimpleName();
    static final float CONNECTOR_Z = 1;
    static final float CENTRAL_CHORD_SCALE = 2.2f;
    static final float HALF_STEP_SCALE = 0.7f;
    static final float CHORD_PADDING_DP = 20f;
    private Chord chord;
    private OnChordChangedListener onChordChangedListener;

    TextView selectedChord;
    TextView centralChord;
    ImageView centralChordBackground;
    ImageView centralChordThrobber;
    ImageView centralChordTouchPoint;
    TextView halfStepUp;
    TextView halfStepDown;
    ImageView halfStepBackground;
    List<SequenceViews> sequences = new ArrayList<>();

    class SequenceViews {
        final ImageView axis = inflateAxisView();
        final ImageView connectForward = inflateConnectorView();
        final ImageView connectBack = inflateConnectorView();
        final TextView forward = inflateChordView();
        final TextView back = inflateChordView();
        final ChordSequence sequence;
        SequenceViews(final ChordSequence s) {
            this.sequence = s;
            forward.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(!forward.getText().equals(centralChord.getText())) {
                        selectedChord = (TextView) v;
                        setChord(sequence.forward(chord));
                    }
                }
            });
            forward.setText(sequence.forward(chord).getName());
            back.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    if(!back.getText().equals(centralChord.getText())) {
                        selectedChord = (TextView) v;
                        setChord(sequence.back(chord));
                    }
                }
            });
            back.setText(sequence.back(chord).getName());
        }
    }
    public interface OnChordChangedListener {
        void onChordChanged(Chord c);
    }


    public TopologyView(Context context) {
        super(context);
        init();
    }

    public TopologyView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public TopologyView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void setOnChordChangedListener(OnChordChangedListener onChordChangedListener) {
        this.onChordChangedListener = onChordChangedListener;
    }

    private void init() {
        chord = new Chord(0, MAJ_7);
        centralChord = inflateChordView(7);
        halfStepUp = inflateChordView(4);
        halfStepDown = inflateChordView(4);
        halfStepUp.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                selectedChord = (TextView) v;
                setChord(CHROMATIC.forward(chord));
            }
        });
        halfStepDown.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                selectedChord = (TextView) v;
                setChord(CHROMATIC.back(chord));
            }
        });
        inflateBG();
        updateChordText();
        post(new Runnable() {
            @Override
            public void run() {
                NavigationAnimations.skipToInitialState(TopologyView.this);
                NavigationAnimations.animateToSelectionPhase(TopologyView.this);
            }
        });
    }

    public Chord getChord() {
        return chord;
    }

    public void setChord(Chord c) {
        chord = c;
        if(selectedChord != null) {
            NavigationAnimations.animateToTargetChord(this);
        } else {
            updateChordText();
        }
        if(onChordChangedListener != null) onChordChangedListener.onChordChanged(chord);
    }

    void updateChordText() {
        centralChord.setText(chord.getName());
        halfStepUp.setText(CHROMATIC.forward(chord).getName());
        halfStepDown.setText(CHROMATIC.back(chord).getName());
        for(SequenceViews sv : sequences) {
            sv.forward.setText(sv.sequence.forward(chord).getName());
            sv.back.setText(sv.sequence.back(chord).getName());
        }
    }

    private TextView inflateChordView() {
        return inflateChordView(2);
    }

    private TextView inflateChordView(float defaultZ) {
        LayoutInflater.from(getContext()).inflate(R.layout.topology_chord, this, true);
        TextView result = (TextView) findViewWithTag("newChord");
        result.setTag(null);
        result.setZ(defaultZ);
        return result;
    }

    private ImageView inflateAxisView() {
        LayoutInflater.from(getContext()).inflate(R.layout.topology_axis, this, true);
        ImageView result = (ImageView) findViewWithTag("newConnector");
        result.setZ(0);
        result.setTag(null);
        return result;
    }

    private ImageView inflateConnectorView() {
        LayoutInflater.from(getContext()).inflate(R.layout.topology_connector, this, true);
        ImageView result = (ImageView) findViewWithTag("newConnector");
        result.setZ(CONNECTOR_Z);
        result.setTag(null);
        return result;
    }

    private void inflateBG() {
        LayoutInflater.from(getContext()).inflate(R.layout.topology_bg, this, true);
        centralChordBackground = (ImageView) findViewWithTag("newBG");
        centralChordBackground.setZ(5);
        centralChordBackground.setTag(null);

        LayoutInflater.from(getContext()).inflate(R.layout.topology_bg_half_steps, this, true);
        halfStepBackground = (ImageView) findViewWithTag("newBG");
        halfStepBackground.setZ(3);
        halfStepBackground.setTag(null);

        LayoutInflater.from(getContext()).inflate(R.layout.topology_bg_highlight, this, true);
        centralChordThrobber =  (ImageView) findViewWithTag("newBG");
        centralChordThrobber.setZ(6);
        centralChordThrobber.setTag(null);
        centralChordThrobber.setAlpha(0f);

        LayoutInflater.from(getContext()).inflate(R.layout.topology_bg, this, true);
        centralChordTouchPoint = (ImageView) findViewWithTag("newBG");
        centralChordTouchPoint.setZ(1000);
        centralChordTouchPoint.setAlpha(0f);
        centralChordTouchPoint.setTag(null);
    }

    public void addSequence(int index, ChordSequence sequence) {
        if(!containsSequence(sequence)) {
            sequences.add(index, new SequenceViews(sequence));
            NavigationAnimations.animateToSelectionPhase(this);
        }
    }

    public void addSequence(ChordSequence sequence) {
        addSequence(sequences.size(), sequence);
    }

    public void removeSequence(ChordSequence sequence) {
        for(int index = 0; index < sequences.size(); index++) {
            SequenceViews views = sequences.get(index);
            if(views.sequence == sequence) {
                animateViewOut(views.axis);
                animateViewOut(views.forward);
                animateViewOut(views.back);
                animateViewOut(views.connectForward);
                animateViewOut(views.connectBack);
                sequences.remove(index);
                NavigationAnimations.animateToSelectionPhase(this);
                break;
            }
        }
    }

    private void animateViewOut(final View child) {
        child.animate().alpha(0f).withEndAction(new Runnable() {
            @Override
            public void run() {
                TopologyView.this.removeView(child);
            }
        });
    }

    public boolean containsSequence(ChordSequence sequence) {
        for(int index = 0; index < sequences.size(); index++) {
            if(sequences.get(index).sequence == sequence) {
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/kotlin/com/jonlatane/beatpad/view/topology/TopologyView.kt
package com.jonlatane.beatpad.view.topology

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import com.jonlatane.beatpad.R
import com.jonlatane.beatpad.harmony.ChordSequence
import com.jonlatane.beatpad.harmony.Topology
import com.jonlatane.beatpad.harmony.chord.Chord
import com.jonlatane.beatpad.harmony.chord.Maj7
import com.jonlatane.beatpad.harmony.chordsequence.Chromatic
import com.jonlatane.beatpad.showTopologyPicker
import org.jetbrains.anko.*
import org.jetbrains.anko.sdk25.coroutines.onClick
import java.util.*
import kotlin.properties.Delegates.observable

class TopologyView @JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null,
	defStyle: Int = 0
) : RelativeLayout(context, attrs, defStyle) {

	inline fun <T: View> T.lparams(
		width: Int = android.view.ViewGroup.LayoutParams.WRAP_CONTENT,
		height: Int = android.view.ViewGroup.LayoutParams.WRAP_CONTENT,
		init: RelativeLayout.LayoutParams.() -> Unit
	): T {
		val layoutParams = RelativeLayout.LayoutParams(width, height)
		layoutParams.init()
		this@lparams.layoutParams = layoutParams
		return this
	}
	var onChordChangedListener: ((Chord) -> Unit)? by observable<((Chord) -> Unit)?>(null) {
		_, _, listener ->
		listener?.invoke(chord)
	}
	var chord: Chord by observable(Chord(0, Maj7)) { _, _, chord ->
		if (selectedChord != null) {
			animateTo(InitialState)
		} else {
			updateChordText()
		}
		onChordChangedListener?.invoke(chord)
	}
	var topology: Topology by observable(Topology.intermediate) {
		_, _, new ->
		(Topology.allSequences - new).forEach {
			removeSequence(it)
		}
		new.indices.forEach {
			addSequence(it, new[it])
		}
	}
	internal var centralChord: TextView
	internal lateinit var centralChordThrobber: View
	internal lateinit var centralChordTouchPoint: View
	internal var halfStepUp: TextView
	internal var halfStepDown: TextView
	internal lateinit var halfStepBackground: View
	internal var selectedChord: TextView? = null
	internal var sequences: MutableList<SequenceViews> = ArrayList()

	internal inner class SequenceViews(val sequence: ChordSequence) {
		val axis = inflateAxisView()
		val connectForward = inflateConnectorView()
		val connectBack = inflateConnectorView()
		val forward = inflateChordView()
		val back = inflateChordView()

		init {
			forward.setOnClickListener { v: View ->
				if (forward.text != centralChord.text) {
					selectedChord = v as TextView
					chord = sequence.forward(chord)
				}
			}
			forward.text = sequence.forward(chord).name
			back.setOnClickListener { v: View ->
				if (back.text != centralChord.text) {
					selectedChord = v as TextView
					chord = sequence.back(chord)
				}
			}
			back.text = sequence.back(chord).name
		}
	}

	init {
		clipToPadding = false
		clipChildren = false
		centralChord = inflateChordView(centralChordElevation)
		halfStepUp = inflateChordView(halfStepChordElevation)
		halfStepDown = inflateChordView(halfStepChordElevation)
		halfStepUp.setOnClickListener { v: View ->
			selectedChord = v as TextView
			chord = Chromatic.forward(chord)
		}
		halfStepDown.setOnClickListener { v: View ->
			selectedChord = v as TextView
			chord = Chromatic.back(chord)
		}
		inflateBG()
		updateChordText()
		post {
			skipTo(InitialState)
			this.topology = topology
			animateTo(SelectionState)
		}
		val topology = this
		button {
			text = "Mode"
			onClick {
				showTopologyPicker(topology)
			}
		}.lparams {
			alignParentBottom()
			alignParentRight()
			elevation = 1f
		}

		backgroundColor = resources.getColor(R.color.colorPrimaryDark)
	}

	internal fun updateChordText() {
		fun TextView.chordify(c: Chord) {
			text = c.name
			val p = arrayOf(paddingLeft, paddingTop, paddingRight, paddingBottom)
			backgroundResource = when {
				c.isDominant -> R.drawable.topology_chord_dominant
				c.isDiminished -> R.drawable.topology_chord_diminished
				c.isMinor -> R.drawable.topology_chord_minor
				c.isAugmented -> R.drawable.topology_chord_augmented
				c.isMajor -> R.drawable.topology_chord_major
				else -> R.drawable.topology_chord
			}
			setPadding(p[0], p[1], p[2], p[3])
		}

		centralChord.chordify(chord)
		halfStepUp.chordify(Chromatic.forward(chord))
		halfStepDown.chordify(Chromatic.back(chord))
		for (sv in sequences) {
			sv.forward.chordify(sv.sequence.forward(chord))
			sv.back.chordify(sv.sequence.back(chord))
		}
	}

	private fun inflateChordView(defaultElevation: Float = defaultChordElevation): TextView {
		LayoutInflater.from(context).inflate(R.layout.topology_chord, this, true)
		val result = findViewWithTag("newChord").apply {
			elevation = defaultElevation
			tag = null
		} as TextView
		return result
	}

	private fun inflateAxisView(): View {
		LayoutInflater.from(context).inflate(R.layout.topology_axis, this, true)
		val result = findViewWithTag("newConnector").apply {
			elevation = axisElevation
			tag = null
		}
		return result
	}

	private fun inflateConnectorView(): View {
		LayoutInflater.from(context).inflate(R.layout.topology_connector, this, true)
		val result = findViewWithTag("newConnector").apply {
			elevation = connectorElevation
			tag = null
		}
		return result
	}

	private fun inflateBG() {
		LayoutInflater.from(context).inflate(R.layout.topology_bg_half_steps, this, true)
		halfStepBackground = findViewWithTag("newBG").apply {
			elevation = halfStepBackgroundElevation
			tag = null
		}

		LayoutInflater.from(context).inflate(R.layout.topology_bg_highlight, this, true)
		centralChordThrobber = findViewWithTag("newBG").apply {
			outlineProvider = null
			elevation = Float.MAX_VALUE - 1f
			tag = null
			alpha = 0f
		}

		LayoutInflater.from(context).inflate(R.layout.topology_bg_highlight, this, true)
		centralChordTouchPoint = findViewWithTag("newBG").apply {
			elevation = Float.MAX_VALUE
			alpha = 0f
			tag = null
		}
	}

	fun onResume() {
		animateTo(SelectionState)
	}

	private fun addSequence(index: Int, sequence: ChordSequence) {
		if (!containsSequence(sequence)) {
			sequences.add(index, SequenceViews(sequence))
			updateChordText()
			post { animateTo(SelectionState) }
		}
	}

	private fun removeSequence(sequence: ChordSequence) {
		for (index in 0..sequences.size - 1) {
			val views = sequences[index]
			if (views.sequence === sequence) {
				animateViewOut(views.axis)
				animateViewOut(views.forward)
				animateViewOut(views.back)
				animateViewOut(views.connectForward)
				animateViewOut(views.connectBack)
				sequences.removeAt(index)
				animateTo(SelectionState)
				break
			}
		}
	}

	private fun animateViewOut(child: View) {
		child.animate().alpha(0f).withEndAction {
			this@TopologyView.removeView(child)
		}
	}

	fun containsSequence(sequence: ChordSequence): Boolean {
		return (0..sequences.size - 1).any { sequences[it].sequence === sequence }
	}

	fun skipTo(state: NavigationState) = state.skipTo(this)
	fun animateTo(state: NavigationState) = state.animateTo(this)
}
